### ISSUE TITLE
metric: changes in histogram buckets affects the connection latency graph in db console

### ISSUE THREAD

**Issue by ZhouXing19** [cockroachdb/cockroach#95833]
While investigating https://github.com/cockroachdb/cockroach/issues/95627, we noticed that the p90 connection latency from a local client to a server on roachprod jumped from ~230ms to ~500ms since https://github.com/cockroachdb/cockroach/commit/a82aa8267b1b33a685a02a0f73ec2ece1f8ba295. 

Before:
<img width="955" alt="Screenshot 2023-01-24 at 18 42 26" src="https://user-images.githubusercontent.com/53610260/214606632-137148d3-b6d9-420b-baa9-790c6e66f632.png">

After:
<img width="970" alt="Screenshot 2023-01-24 at 18 43 14" src="https://user-images.githubusercontent.com/53610260/214606682-b40a2131-bf2a-4759-a985-be897f8f4ef9.png">

While the client latency (i.e. the time for the query to finish from the client side) are the same, ~500ms. 

**How to repro**
1. Build the Linux binaries before and after the commit https://github.com/cockroachdb/cockroach/commit/a82aa8267b1b33a685a02a0f73ec2ece1f8ba295 (say they are `cockroach-pre` and `cockroach-post`).
2. Create 2 roachprod (i.e. `$USER-test-pre` and `$USER-test-post`)
3. Use [this script](https://gist.github.com/ZhouXing19/360eab6168ec15b42896a4c72835a870#file-prepare-sh) for the testing. (i.e. run `./prepare.sh pre` or `./prepare.sh post`). Note that you'll need to create the `add_user.sql` as well, as specified [here](https://gist.github.com/ZhouXing19/360eab6168ec15b42896a4c72835a870#file-prepare-sh-L27-L32).
4. Once the connection start to establish (i.e. you can see `----- round 1 -----` from the terminal stdout), go to [the db console](https://gist.github.com/ZhouXing19/360eab6168ec15b42896a4c72835a870#file-prepare-sh-L43) and check the `connection latency` under the `dist SQL` section.  

Jira issue: CRDB-23769

**Comment by thtruo** [comment]
cc @abarganier assigning to you as you're currently investigating it

**Comment by abarganier** [comment]
Okay, here's the scoop.

This is caused by switching from the older `hdrhistogram`, which dynamically created its own buckets based on maxVal and significant figures parameters, to Prometheus-based histograms, where the histogram buckets ad explicitly defined, such as: https://github.com/cockroachdb/cockroach/blob/master/pkg/util/metric/histogram_buckets.go#L17-L34

These buckets are not granular enough, and their value ranges don't make sense for measuring SQL latencies. This is the cause of the misleading latency values - we have less precision and therefore these ~200ms latencies are ending up in the ~517ms bucket.

We're going to backport a revert of https://github.com/cockroachdb/cockroach/pull/86671 into 22.2. This allows us to maintain the code for the newer histogram while reverting the metrics themselves to use the older histograms.

This will allow us to fix the issue as fast as possible for customers without derailing the team's other initiatives. Once we finish delivering on our outstanding epics for 23.1, we can do a proper fix. A proper fix involves A) defining more appropriate & granular histogram buckets, and B) creating a cluster setting that allows users to optionally enable the new histograms (disabled by default).

We'll also be testing out the impact on continuity here. If customers have been using Prometheus on v22.2 for the past month (for example), what's the impact on Prometheus of the buckets suddenly changing back to the older form? We seek to define that alongside our fix. 

We'll also provide a technical advisory and documentation for TSE on this issue.

**Comment by irfansharif** [comment]
> This allows us to maintain the code for the newer histogram while reverting the metrics themselves to use the older histograms.

Will the revert undo the benefits of https://github.com/cockroachdb/cockroach/pull/85990, which drastically reduced the bucket count for most histograms? The wholesale revert, to me, feels very heavy handed, and especially for a backport, given what we've spelled out for proper fixes (defining more appropriate bucket boundaries, perhaps even narrowly targeting the connection latency graph). 

Above you linked to the IO latency buckets, but for this [connection latency histogram](https://github.com/cockroachdb/cockroach/blob/62422f89fccbf8ba4f43201b474cd598302f7753/pkg/ccl/sqlproxyccl/metrics.go#L216-L220), aren't we using the network latency buckets defined here? 
https://github.com/cockroachdb/cockroach/blob/a82aa8267b1b33a685a02a0f73ec2ece1f8ba295/pkg/util/metric/histogram_buckets.go#L40-L57

So we have bucket boundaries  [196.171733ms, 337.616984ms, 581.048177ms], which when computing various percentiles, we interpolate within: https://github.com/cockroachdb/cockroach/blob/4a2d06a293412f62f5f05a3b0b799d30080592e7/pkg/util/metric/metric.go#L665-L674

So I'm a bit confused about how, with those bucket boundaries and that linear interpolation code, we could have "~200ms latencies [..] ending up in the ~517ms bucket".

**Comment by irfansharif** [comment]
Oh I was looking at the wrong thing, sorry for the noise. We are indeed using the IOLatencyBuckets here. 
https://github.com/cockroachdb/cockroach/blob/2f811420a1492ca547659003f193233341b7fd87/pkg/sql/pgwire/server.go#L276-L278 where the bucket boundaries are further apart around ~200ms ([193.069772ms, 517.947467ms]).

**Comment by abarganier** [comment]
I don't disagree that it's a bit heavy handed of an approach. However, at the same time, I don't feel comfortable defining new sets of buckets to use without an accompanying cluster setting to disable using the new histograms altogether. The benefits of reduced #'s of buckets is negated if our histograms are misleading and no longer useful. 
 
Unfortunately, the team does not have capacity right now to put our other initiatives on the shelf to introduce a new cluster setting and the ability to toggle back and forth between the old and new histograms. Given the prevalence of customer complaints, the priority for now is to make histograms useful again, so the team has opted to take the path of least resistance to get that done as soon as possible. We could attempt to fix the buckets by defining new ones, but what if those are also found to have issues? Then we're back to square one and we find ourselves in the same situation, which is really bad optics when it comes to our customers who have widely noticed this issue. This is why I don't feel comfortable with this solution unless it's accompanied by a cluster setting to toggle between the new & old histograms, and that's a non-trivial change AFAIK.

Reintroducing the new histograms with an accompanying cluster setting will be fairly high priority for the team, but we have bigger commitments on our plate right now. However, I expect we'll be able to get it in before the 23.1 branch cut (along with a backport of the fix to 22.2), so keep in mind that this is only temporary.

**Comment by abarganier** [comment]
Hey all - the team has discussed this at length and we've decided against the revert. We believe that it will create toil for customers, and that we can provide a better solution in time for the upcoming v22.2.4 patch release.

Here's the revised plan:

- Provide a cluster setting that enables customers to disable the new histogram buckets in favor of the legacy hdr histograms. This will require a node restart to take effect, but provides a way to revert back to the older histograms if issues arise with the newer histograms and/or their pre-defined buckets.

- For the new histograms, update all pre-defined bucket types with a significant increase in granularity. Right now, each pre-defined bucket type only has 15 buckets. We will be increasing this to 60 for each of them, which should increase the fidelity significantly. If folks feel that a bucket should be even higher fidelity, we can work that out in the PR review. We are open to feedback on how to best serve different use cases.

@dhartunian and I have already made good progress on this approach. We will continue posting updates here as we have them.
